### PR TITLE
Support arrays of scalars inside structured types

### PR DESCRIPTION
**Pull request by gmarkall**
This adds support for arrays of scalars inside structured types by represented these arrays as a `types.VoidArray`, which behaves similarly to a `types.Array`, but the shape rather than just the number of dimensions is part of the type.

Only the raw array data is stored in a structured type, so an array structure is created and populated with relevant information when an array inside a structured type is accessed.

### CODE REVIEW

File: numba/types.py [hunk 1]
```diff
@@ -714,6 +715,33 @@ def is_contig(self):
         return self.layout in 'CF'
 
 
+class VoidArray(Array):
+    """
+    A VoidArray is an array nested within a structured type (which are "void"
+    type in NumPy parlance). Unlike an Array, the shape, and not just the number
+    of dimenions is part of the type of a VoidArray.
```
Comment (pitrou) [inline numba/types.py]: Typo in "dimensions"

File: numba/types.py [hunk 2]
```diff
@@ -714,6 +715,33 @@ def is_contig(self):
         return self.layout in 'CF'
 
 
+class VoidArray(Array):
+    """
+    A VoidArray is an array nested within a structured type (which are "void"
```
Comment (pitrou) [inline numba/types.py]: It would be less obscure if named e.g. NestedArray or SubArray.

File: numba/types.py [hunk 3]
```diff
@@ -714,6 +715,33 @@ def is_contig(self):
...
+    type in NumPy parlance). Unlike an Array, the shape, and not just the number
+    of dimenions is part of the type of a VoidArray.
+    """
+
+    def __init__(self, dtype, shape, const=False):
```
Comment (pitrou) [inline numba/types.py]: Is const-ness ever used here?

File: numba/types.py (hunk 1 above)
Comment (gmarkall) [inline numba/types.py]: Thanks - will fix

File: numba/types.py (hunk 2 above)
Comment (gmarkall) [inline numba/types.py]: OK - renamed to `NestedArray`... I had been trying to keep consistency with NumPy, but I agree that that actually makes things less clear.

**Comment by gmarkall** [comment]
Looks like I need to do something to make this work correctly with the recent data model changes.

**Comment by gmarkall** [comment]
I think this is fixed up now - using the `ArrayModel` for the `NestedArray` seems to work because we create an `ArrayTemplate`-like struct for it when it is accessed.

File: numba/targets/base.py [hunk 4]
```diff
@@ -437,11 +437,40 @@ def get_attribute(self, val, typ, attr):
...
+                    size = dtype.bitwidth//8
+                    ary.itemsize = context.get_constant(types.intp, size)
+                    dataty = self.get_data_type(dtype)
+                    ary.data = cgutils.get_record_member(builder, val, offset,
+                                                         dataty)
+                    ary.shape = cgutils.pack_array(builder,
+                                                   [ self.get_constant(types.intp, s)
+                                                     for s in elemty.shape ])
+                    stride = size
+                    strides = []
```
Comment (pitrou) [inline numba/targets/base.py]: Since shape and nitems are already defined on NestedArray, perhaps put strides and size there too?

Comment (gmarkall) [inline numba/targets/base.py]: OK - moved them over to `NestedArray`. I don't think the size will vary depending on the target type for the present use case.

**Comment by pitrou** [comment]
Merging, thank you!

**Pull request merged**